src/broadcast.rs

```diff
@@ -89,19 +86,68 @@ pub struct Broadcast<NodeUid> {
     /// Whether we have already output a value.
     has_output: bool,
     /// The proofs we have received via `Echo` messages, by sender ID.
-    echos: BTreeMap<NodeUid, Proof<Vec<u8>>>,
+    echos: BTreeMap<N, Proof<Vec<u8>>>,
     /// The root hashes we received via `Ready` messages, by sender ID.
-    readys: BTreeMap<NodeUid, Vec<u8>>,
+    readys: BTreeMap<N, Vec<u8>>,
+    /// The outgoing message queue.
+    messages: VecDeque<TargetedMessage<BroadcastMessage, N>>,
+    /// The output, if any.
+    output: Option<Vec<u8>>,
 }
 
-impl<NodeUid: Eq + Hash + Debug + Clone + Ord> Broadcast<NodeUid> {
+impl<N: Eq + Debug + Clone + Ord> DistAlgorithm for Broadcast<N> {
+    type NodeUid = N;
+    type Input = Vec<u8>; // TODO: Allow anything serializable.
+    type Output = Self::Input;
+    type Message = BroadcastMessage;
+    type Error = Error;
+
+    fn input(&mut self, input: Self::Input) -> Result<(), Self::Error> {
+        if self.our_id != self.proposer_id {
+            return Err(Error::UnexpectedMessage);
+        }
+        // Split the value into chunks/shards, encode them with erasure codes.
+        // Assemble a Merkle tree from data and parity shards. Take all proofs
+        // from this tree and send them, each to its own node.
+        let proof = self.send_shards(input)?;
+        // TODO: We'd actually need to return the output here, if it was only one node. Should that
+        // use-case be supported?
+        let our_id = self.our_id.clone();
+        self.handle_value(&our_id, proof)
+    }
+
+    fn handle_message(&mut self, sender_id: &N, message: Self::Message) -> Result<(), Self::Error> {
+        if !self.all_uids.contains(sender_id) {
+            return Err(Error::UnknownSender);
+        }
+        match message {
+            BroadcastMessage::Value(p) => self.handle_value(sender_id, p),
+            BroadcastMessage::Echo(p) => self.handle_echo(sender_id, p),
+            BroadcastMessage::Ready(ref hash) => self.handle_ready(sender_id, hash),
+        }
+    }
+
+    fn next_message(&mut self) -> Option<TargetedMessage<Self::Message, N>> {
+        self.messages.pop_front()
+    }
+
+    fn next_output(&mut self) -> Option<Self::Output> {
```

Comment:
In what sense does this provide a "next" output given that there should be at most one output? I think of the output more in terms of generators: https://doc.rust-lang.org/1.23.0/std/ops/trait.Generator.html
I think this function is performing a `Return` action of a generator `resume`.